sys/arm/arm/generic_timer.c

```diff
@@ -292,15 +292,24 @@ cntpct_handler(vm_offset_t va, uint32_t insn, struct trapframe *frame,
 	if ((insn & MRS_MASK) != MRS_VALUE)
 		return (0);
 
-	if (MRS_SPECIAL(insn) != MRS_SPECIAL(CNTPCT_EL0))
+	if (MRS_SPECIAL(insn) == MRS_SPECIAL(CNTPCT_EL0)) {
```

Comment:
Hmm, I would maybe restructure this to avoid duplicating quite as much code, maybe something like:

```
switch (MSR_SPECIAL(insn)) {
case MSR_SPECIAL(CNTPCT_EL0):
        val = READ_SPECIALREG(cntvct_el0);
        break;
case MSR_SPECIAL(CNTFREQ_EL0):
        val = READ_SPECIALREG(cntfreq_el0);
        break;
default:
        return (0);
}

reg = MSR_REGISTER(insn);
...
```

However, re-reading this now, this does point out that the emulation of CNTPCT is actually reading the value of CNTVCT (not CNTPCT) and returning it to userland (even more reason to not permit the vdso to use this at all).

Reply:
I wonder at this point whether we want to just yank out of the vdso code the use of cntpct since it seems to be never used anyway. This would be an upstream change, I guess.